### PR TITLE
Add Network Address Translation (NAT) for NuttX

### PR DESCRIPTION
**Pull request by wengzhe**
## Summary
Add full cone NAT for NuttX, supports TCP / UDP / ICMP ECHO (PING)

Outbound: LAN ->  Forward  ->  NAT(only if targeting at WAN)  -> WAN
Inbound:  WAN ->  NAT(only from WAN, change dest) -> Forward  -> LAN

patches included:
- net: add basic NAT workflow
- net: verify NAT port usage in tcp_selectport
- net: select NAT external port by tcp_selectport for TCP
- net/nat: Add TCP entry expiration logic
- net/nat: Add ICMP ECHO (REQUEST & REPLY) support
- net/nat: Add UDP support
- Documentation: add NAT description

## Impact
If `CONFIG_NET_NAT` is enabled, and `ipv4_nat_enable` is called on any netdev (which sets `IFF_NAT`), the device will be regarded as WAN and outbound packets will be masqueraded on the device.

## Testing
Tested on Ubuntu 22.04 x86_64 by following steps:
1. Configure NuttX with >=2 TAP devices (host route mode) and NAT enabled:
```Kconfig
CONFIG_NET_IPFORWARD=y
CONFIG_NET_NAT=y
# CONFIG_SIM_NET_BRIDGE is not set
CONFIG_SIM_NETDEV_NUMBER=2
```
2. Call `ipv4_nat_enable` on one dev on startup
```C
/* arch/sim/src/sim/up_netdriver.c */
int netdriver_init(void)
{
  ...
  ipv4_nat_enable(&g_sim_dev[0]);
  ...
}
```
3. Set IP Address for NuttX on startup
```shell
ifconfig eth0 10.0.1.2
ifup eth0
ifconfig eth1 10.0.10.2
ifup eth1
```
4. Configure IP & namespace & route on host side (maybe need to be root, then try `sudo -i`)
```bash
IF_HOST="enp1s0"
IF_0="tap0"
IP_HOST_0="10.0.1.1"
IF_1="tap1"
IP_HOST_1="10.0.10.1"
IP_NUTTX_1="10.0.10.2"

# add net namespace LAN for $IF_1
ip netns add LAN
ip netns exec LAN sysctl -w net.ipv4.ip_forward=1
ip link set $IF_1 netns LAN
ip netns exec LAN ip link set $IF_1 up
ip netns exec LAN ip link set lo up

# add address and set default route
ip addr add $IP_HOST_0/24 dev $IF_0
ip netns exec LAN ip addr add $IP_HOST_1/24 dev $IF_1
ip netns exec LAN ip route add default dev $IF_1 via $IP_NUTTX_1

# nat to allow NuttX to access the internet
iptables -t nat -A POSTROUTING -o $IF_HOST -j MASQUERADE
iptables -A FORWARD -i $IF_HOST -o $IF_0 -j ACCEPT
iptables -A FORWARD -i $IF_0 -o $IF_HOST -j ACCEPT
sysctl -w net.ipv4.ip_forward=1
```
5. Do anything in the LAN namespace will go through NAT
```shell
# Host side
iperf -B 10.0.1.1 -s -i 1
# LAN side
sudo ip netns exec LAN iperf -B 10.0.10.1 -c 10.0.1.1 -i 1
```
```shell
# Host side
python3 -m http.server
# LAN side
for i in {1..20000}; do sudo ip netns exec LAN curl 'http://10.0.1.1:8000/' > /dev/null 2>1; done
```
```shell
# LAN side
sudo ip netns exec LAN ping 8.8.8.8
```
```shell
# Host side
tcpdump -nn -i tap0
# LAN side
sudo ip netns exec LAN tcpdump -nn -i tap1
```

### CODE REVIEW

**Review by acassis — CHANGES_REQUESTED** [review]
Please transform this Testing steps in a Documentation/ to let please understand and duplicate it.

File: net/nat/ipv4_nat.c [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+{
+  uint16_t iphdrlen = (ipv4->vhl & IPv4_HLMASK) << 2;
+  FAR struct tcp_hdr_s *tcp =
+      (FAR struct tcp_hdr_s *)((FAR uint8_t *)ipv4 + iphdrlen);
+  FAR struct ipv4_nat_entry *entry = ipv4_nat_outbound_entry_find(
+      IP_PROTO_TCP, net_ip4addr_conv32(ipv4->srcipaddr), tcp->srcport);
+  if (!entry)
+    {
+      /* Outbound entry creation failed, should have corresponding entry. */
+
```
Comment (acassis) [inline net/nat/ipv4_nat.c:256]: Should it be reported using nwarn or ninfo?

File: net/nat/ipv4_nat.c [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+ *
+ * TODO:
+ *   Support multiple NAT devices with isolated external port mapping.
+ ****************************************************************************/
+
+int ipv4_nat_enable(FAR struct net_driver_s *dev)
+{
+  if (IFF_IS_NAT(dev->d_flags))
+    {
+      return -EEXIST;
```
Comment (acassis) [inline net/nat/ipv4_nat.c:429]: nwarn reporting that NAT was already enabled

File: net/nat/ipv4_nat.c [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+ *   Zero is returned if NAT function is successfully disabled on the device;
+ *   A negated errno value is returned if failed.
+ *
+ ****************************************************************************/
+
+int ipv4_nat_disable(FAR struct net_driver_s *dev)
+{
+  if (!IFF_IS_NAT(dev->d_flags))
+    {
+      return -ENODEV;
```
Comment (acassis) [inline net/nat/ipv4_nat.c:455]: nwarn reporting that NAT was not enabled for this interface

File: net/nat/ipv4_nat.c (hunk 2 above)
Comment (wengzhe) [reply net/nat/ipv4_nat.c:431]: Done.

File: net/nat/ipv4_nat.c (hunk 3 above)
Comment (wengzhe) [reply net/nat/ipv4_nat.c:458]: Done.

**Comment by wengzhe** [comment]
Oops, when I try to re-request a review from @acassis, github removed request for @pkarashchenko and I don't know how to add back again.

**Review by xiaoxiang781216 — APPROVED** [review]

File: net/nat/ipv4_nat.c (hunk 1 above)
Comment (acassis) [reply net/nat/ipv4_nat.c:257]: Ok

**Review by acassis — APPROVED** [review]
It should be nice to put it under a "Network" that doesn't exist yet, but it can be done later

File: net/nat/ipv4_nat.c [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Adjust checksums in headers. */
+
+#define chksum_adjust(chksum,old_data,new_data) \
+  net_chksum_adjust((FAR uint16_t *)&chksum, \
+                    (FAR uint16_t *)&old_data, sizeof(old_data), \
+                    (FAR uint16_t *)&new_data, sizeof(new_data))
```
Comment (pkarashchenko) [inline net/nat/ipv4_nat.c:50]:
```suggestion
  net_chksum_adjust((FAR uint16_t *)&(chksum), \
                    (FAR uint16_t *)&(old_data), sizeof(old_data), \
                    (FAR uint16_t *)&(new_data), sizeof(new_data))
```

**Review by pkarashchenko — COMMENTED** [review]
In general looks good.

File: net/nat/ipv4_nat.c (hunk 4 above)
Comment (wengzhe) [reply net/nat/ipv4_nat.c]: Done.

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**